Ear 1113 refactor comments

MattElcock opened the pull request:
### What is the context of this PR?

This PR refactors the comments panel.

I have:

* Created two new components: Comment, and CommentEditor.
* Moved the root CommentPanel into App from Components.

### How to review

Make sure everything is still working as it did before.

- Import the **Capability examples** questionnaire from pre-prod Author into your local environment and:
  - ensure it can be opened in Author;
  - then, ensure it can be viewed in Runner by pressing the **view survey** button
- Does this require a migration? We need one if existing JSON schema properties change

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

tjamesmac commented on eq-author/src/components/Collapsible/index.js:
```diff
@@ -81,42 +81,86 @@ const HideThisButton = styled(Button)`
+}) => {
   const [isOpen, setIsOpen] = useState(defaultOpen);
 
+  const renderTitle = (showHide, isOpen, title) => {
+    if (showHide && isOpen) {
+      return `Hide ${title}`;
+    } else if (showHide && !isOpen) {
+      return `Show ${title}`;
+    }
+
+    return title;
+  };
```
Can move this function outside the component because all of the arguments are being passed in

tjamesmac commented on eq-author/src/components/Comments/Comment/index.js:
```diff
@@ -0,0 +1,225 @@
+
+  return (
+    <>
+      {numOfReplies === 0 && <React.Fragment />}
```
not sure what this is doing

MattElcock replied:
If there's no replies, it chucks out a fragment instead to not display the replies toggle. But, I could instead swap it to just return nothing from the component if there's no replies. I'll take a look at this.

tjamesmac commented on eq-author/src/components/Comments/Comment/index.js:
```diff
@@ -0,0 +1,225 @@
+  const [addReplyVisible, showAddReply] = useState(false);
+  const [replyBtnVisible, showReplyBtn] = useState(!disableReplies);
+
+  const ReplyBtn = styled(Button)`
+    margin-bottom: 1em;
+  `;
```
define outside the component

tjamesmac commented on eq-author/src/components/Comments/Comment/index.js:
```diff
@@ -0,0 +1,225 @@
+        >
+          <ul>
+            {replies.map(({ id, ...rest }) => (
+              <li key={`list-item-comment-${id}`}>
```
is the `<li>` outside the `<Comment>` because it might be a single entity at some point?

MattElcock replied:
Yeah.

tjamesmac commented on eq-author/src/components/Comments/Comment/index.test.js:
```diff
@@ -0,0 +1,155 @@
+  });
+
+  it("Can render", () => {
+    const { getByTestId } = render(
+      <Comment
+        comment={mockComment}
+        replies={mockReplies}
+        onUpdateComment={mockOnUpdateComment}
+        onDeleteComment={mockOnDeleteComment}
+        onAddReply={mockOnAddReply}
+        onUpdateReply={mockOnUpdateReply}
+        onDeleteReply={mockOnDeleteReply}
+      />
+    );
```
could make this into a function so it doesn't have to repeated all the time

tjamesmac commented on eq-author/src/components/Comments/CommentEditor/index.js:
```diff
@@ -0,0 +1,155 @@
+`;
+
+const TextAreaFixedSize = styled.textarea`
+  height: 94px;
+  width: 100%;
+  border: thin solid ${colors.grey};
+  resize: none;
+  font-size: 1em;
+  font-family: inherit;
+  padding: 0.5em;
+  margin-bottom: 0.5em;
+
+  &:focus {
+    ${focusStyle}
+    outline: none;
+  }
+`;
+
+const TextAreaGrowable = styled(TextareaAutosize)`
+  width: 100%;
+  border: thin solid ${colors.grey};
+  resize: none;
+  font-size: 1em;
+  font-family: inherit;
+  padding: 0.5em;
+  margin-bottom: 0.5em;
+
+  &:focus {
+    ${focusStyle}
+    outline: none;
+  }
+`;
```
I think there's one css property difference between these. Could make it into a common css property thing

tjamesmac commented on eq-author/src/components/Comments/CommentEditor/index.test.js:
```diff
@@ -0,0 +1,101 @@
+  });
+
+  it("Can render", () => {
+    const { getByTestId } = render(
+      <CommentEditor onConfirm={mockOnConfirm} onCancel={mockOnCancel} />
+    );
```
Could separate out into a function again with the option for additional props

tjamesmac commented on eq-author/src/components/Comments/PureComment/index.js:
```diff
@@ -0,0 +1,269 @@
+    padding: 0.5em 1em;
+    margin: 0;
+    margin-bottom: 0.3125em;
+  `;
+
+  const Avatar = styled.p`
+    width: 2.25em;
+    height: 2.25em;
+    line-height: 2.25em;
+    border-radius: 50%;
+    text-align: center;
+    color: ${colors.white};
+    background-color: ${colors.primary};
+    margin: 0;
+    margin-right: 0.5em;
+    text-transform: uppercase;
+  `;
+
+  const Author = styled.p`
+    margin: 0;
+    text-transform: capitalize;
+  `;
+
+  const Date = styled.p`
+    font-size: 0.8em;
+    color: ${colors.grey};
+    margin: 0;
+  `;
+
+  const ButtonGroup = styled.div`
+    display: flex;
+
+    button {
+      margin-right: 0.5em;
+    }
+  `;
+
+  const RightButtonGroup = styled(ButtonGroup)`
+    margin-left: auto;
+  `;
```
define outside

tjamesmac commented on eq-author/src/components/Comments/PureComment/index.test.js:
```diff
@@ -0,0 +1,115 @@
+
+  it("Can render", () => {
+    const { getByTestId } = render(
+      <PureComment
+        {...mockComment}
+        onUpdateComment={mockOnUpdateComment}
+        onDeleteComment={mockOnDeleteComment}
+        canDelete
+        canEdit
+      />
```
can separate into a function

tjamesmac commented on eq-author/src/components/Comments/PureComment/index.js:
```diff
@@ -0,0 +1,269 @@
+        </RightButtonGroup>
+      </Header>
+      <Body data-test="PureComment__Body">
+        {!editing && <Text data-test="PureComment__CommentText">{text}</Text>}
+        {editing && (
+          <CommentEditor
+            data-test="PureComment__CommentEditor"
+            showCancel
+            confirmText={"Save"}
+            replyId={replyId}
+            commentId={commentId}
+            initialValue={text}
+            variant={"growable"}
+            onConfirm={(commentText, commentId, replyId) => {
+              onUpdateComment(commentId, commentText, replyId);
+              setEditing(false);
+              showReplyBtn(true);
+            }}
+            onCancel={() => {
+              setEditing(false);
+              showReplyBtn(true);
+            }}
+          />
+        )}
```
```suggestion
        {editing ? (
          <CommentEditor
            data-test="PureComment__CommentEditor"
            showCancel
            confirmText={"Save"}
            replyId={replyId}
            commentId={commentId}
            initialValue={text}
            variant={"growable"}
            onConfirm={(commentText, commentId, replyId) => {
              onUpdateComment(commentId, commentText, replyId);
              setEditing(false);
              showReplyBtn(true);
            }}
            onCancel={() => {
              setEditing(false);
              showReplyBtn(true);
            }}
          />
        ) : <Text data-test="PureComment__CommentText">{text}</Text>}
```

tjamesmac commented on eq-author/src/App/Comments/index.js:
```diff
@@ -0,0 +1,205 @@
+  const formattedComments = comments.map(
+    ({ id, user, commentText, createdTime, editedTime, replies }) => ({
+      id,
+      author: user.displayName || user.name || "",
+      canEdit: user.id === me.id,
+      canDelete: user.id === me.id,
+      datePosted: createdTime,
+      dateModified: editedTime,
+      text: commentText,
+      replies: replies.map(
+        ({ id, user, commentText, createdTime, editedTime }) => ({
+          id,
+          author: user.displayName || user.name || "",
+          canEdit: user.id === me.id,
+          canDelete: user.id === me.id,
+          datePosted: createdTime,
+          dateModified: editedTime,
+          text: commentText,
```
these are mostly duplicated and could be reused

tjamesmac reviewed (changes requested):
There's a few little things I've mentioned in the comments. 

Other than that I haven't fully gone through the front-end yet but:
- the edit and cancel buttons don't have pointer events
- the avatar displayNames are incorrectly capitalised 
![Screenshot 2021-04-07 at 14 56 00](https://user-images.githubusercontent.com/22731314/113878290-6128c700-97b1-11eb-9284-e7e70ab697d8.png)


Just out of curiosity, why do the comment and reply text areas scroll but the edit comment ones grow?

Also how much pain would it be to make the mutations localised to the components that need them versus being at the highest level?

MattElcock commented:
> The edit and cancel buttons don't have pointer events
> The avatar displayNames are incorrectly capitalised

Fixed

> Just out of curiosity, why do the comment and reply text areas scroll but the edit comment ones grow?

Good question! I'm not sure, probably one for Joe. But, that's how it's done currently so I've mimiced that behaviour. I imagine  editing grows in case you're editing a comment that's one line long, as otherwise the box would be super tiny unless you have it start bigger.

> Also how much pain would it be to make the mutations localised to the components that need them versus being at the highest level?

Probably not too painful. The main reason why I didn't do this is so that the component was completely decoupled from the API. I didn't want to have any GraphQL files in the component's directory.

tbaggaley commented:
I would second Tom's request to move the data fetching / mutations to the component that needs them - admittedly a bit more work in that you have to use more `jest.mock(...)` calls in the tests but does avoid unnecessary renders & the necessity to follow prop trails to reverse-engineer stuff if unfamiliar with what it does

farres1 commented:
Hi Matt.
Adding what I mentioned from the meeting. Adding a reply does not automatically close the reply text field, the field can only be closed by leaving the page or clicking cancel.
Clicking reply on a comment when already replying to another comment keeps both reply text fields open. This second part might not be important.

jchapman68 commented:
When using keybard, you can tab to edit a comment but after hitting enter you’re not placed in the edit box but have to tab twice.

MattElcock commented:
> When using keybard, you can tab to edit a comment but after hitting enter you’re not placed in the edit box but have to tab twice.

Thanks Josh. I've fixed this.

MattElcock commented:
> Hi Matt.
> Adding what I mentioned from the meeting. Adding a reply does not automatically close the reply text field, the field can only be closed by leaving the page or clicking cancel.
> Clicking reply on a comment when already replying to another comment keeps both reply text fields open. This second part might not be important.

Cheers Sam. I've fixed the first issue. The second I'll ask around and see what people think as to whether or not it's a bug.

MattElcock commented:
> I would second Tom's request to move the data fetching / mutations to the component that needs them - admittedly a bit more work in that you have to use more `jest.mock(...)` calls in the tests but does avoid unnecessary renders & the necessity to follow prop trails to reverse-engineer stuff if unfamiliar with what it does

Hey Tom, cheers for your feedback. I've actioned these comments; still need to fix the tests and linting, however.

MattElcock commented:
Fixed everything (afaik)! Removed storybook files because it wasn't playing nice with the new GraphQL structure. I'll look at fixing this at a later date, as it's not imperative at this stage.

tjamesmac commented on eq-author/src/App/Comments/index.test.js:
```diff
@@ -0,0 +1,125 @@
+  },
+};
+
+jest.mock("@apollo/react-hooks", () => ({
+  useMutation: () => [mockUseMutation],
+  useQuery: () => mockUseQuery,
+}));
+
+jest.mock("react-apollo", () => ({
+  useSubscription: () => [mockUseSubscription],
+}));
```
```suggestion
jest.mock("@apollo/react-hooks", () => ({
  useMutation: () => [mockUseMutation],
  useQuery: () => mockUseQuery,
    useSubscription: () => [mockUseSubscription],
}));
```
Does this work?

MattElcock replied:
I'll give it a try! Watch this space.

MattElcock replied:
It does. Madness, nice spot. Cheers dude.

tjamesmac reviewed (commented):
We take the editing comments/replies cancel button off?

MattElcock commented:
> We take the editing comments/replies cancel button off?

Sorry! That got removed by accident when I was renaming props. I've added it back in.

tjamesmac reviewed (approved):
Yay! Looks really good, you've also managed to cut like 2000 lines of code out and it's still as understandable

zimboflyman commented on eq-author/src/components/Comments/Comment/index.js:
```diff
@@ -0,0 +1,320 @@
+   * The ID of the comment at the root of the chain.
+   * If the comment is the one at the root of the chain,
+   * the value of this will be equal to ID.
+   */
```
are you happy leaving these descriptive comments in? Most of these propNames are pretty self descriptive

zimboflyman commented on eq-author/src/components/Comments/CommentEditor/index.js:
```diff
@@ -0,0 +1,204 @@
+CommentEditor.propTypes = {
+  /**
+   * The text to show on the confirm button.
+   */
```
comments needed here?
